File: backend/app/api/v1/endpoints/firma_storage.py
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.deps import require_admin_user
from app.core.db import get_db
from app.models.user import User
from app.schemas.firma_storage import FirmaStorageConfigCreate, FirmaStorageConfigResponse
from app.repositories.firma_storage_repository import FirmaStorageRepository
from app.services.storage.factory import StorageFactory

router = APIRouter()

@router.get("/config", response_model=FirmaStorageConfigResponse)
async def get_storage_config(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_admin_user),
):
    """
    Obtiene la configuración de almacenamiento activa para la firma.
    """
    repo = FirmaStorageRepository(db, current_user.firma_id)
    config = await repo.get_config()
    if not config:
        config = await repo.create_or_update_config(
            FirmaStorageConfigCreate(provider="local", auth_type="local")
        )
    return config

@router.post("/config", response_model=FirmaStorageConfigResponse)
async def set_storage_config(
    payload: FirmaStorageConfigCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_admin_user),
):
    """
    Actualiza la configuración de almacenamiento de la firma (Google Drive, OneDrive, Local).
    """
    repo = FirmaStorageRepository(db, current_user.firma_id)
    config = await repo.create_or_update_config(payload)
    return config

@router.post("/config/test")
async def test_storage_connection(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_admin_user),
):
    """
    Realiza un test de conexión en vivo con el proveedor configurado de la firma.
    """
    provider = await StorageFactory.get_provider_for_firma(db, current_user.firma_id)
    try:
        # Intentar crear una carpeta temporal de prueba
        folder_id, url = await provider.create_folder("Test_Conexion_Asuntia")
        repo = FirmaStorageRepository(db, current_user.firma_id)
        config = await repo.get_config()
        if config:
            await repo.update(
                config, {"last_verified_at": datetime.now(timezone.utc)}
            )
        return {
            "status": "success",
            "provider": type(provider).__name__,
            "folder_id": folder_id,
            "web_view_url": url,
            "message": "Conexión con el proveedor probada exitosamente."
        }
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error probando conexión con el proveedor: {str(e)}"
        )

File: frontend/src/features/firma/api/firmaStorage.ts
import { apiClient } from '@/lib/axios';

export interface FirmaStorageConfig {
  id: string;
  firma_id: string;
  provider: 'local' | 'google_drive';
  auth_type: 'local' | 'oauth2';
  root_folder_id?: string;
  root_folder_name?: string;
  is_active: boolean;
  last_verified_at?: string;
}

export async function getGoogleOAuthAuthUrl(): Promise<string> {
  const res = await apiClient.get<{ authorization_url: string }>(
    '/storage/auth-url',
    { params: { provider: 'google_drive' } },
  );
  return res.data.authorization_url;
}

export async function fetchFirmaStorageConfig(): Promise<FirmaStorageConfig> {
  const res = await apiClient.get<FirmaStorageConfig>('/firma/storage/config');
  return res.data;
}

export async function updateFirmaStorageConfig(payload: Partial<FirmaStorageConfig>): Promise<FirmaStorageConfig> {
  const res = await apiClient.post<FirmaStorageConfig>('/firma/storage/config', payload);
  return res.data;
}

export async function testStorageConnection(): Promise<{ status: string; provider: string; folder_id: string; web_view_url: string; message: string }> {
  const res = await apiClient.post<{ status: string; provider: string; folder_id: string; web_view_url: string; message: string }>('/firma/storage/config/test');
  return res.data;
}
